File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0519_Random_Flip_Matrix.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.HashSet;
import java.util.Random;
import java.util.Set;

public class Question_0519_Random_Flip_Matrix {
    private int row;
    private int col;
    private int size;
    private Set<Integer> set;
    private Random random = new Random();

    public Question_0519_Random_Flip_Matrix(int m, int n) {
        this.row = m;
        this.col = n;
        this.size = m * n;
        this.set = new HashSet<>();
    }

    public int[] flip() {
        while (true) {
            int val = random.nextInt(size);
            if (!set.contains(val)) {
                set.add(val);
                return new int[] {val / col, val % col};
            }
        }
    }

    public void reset() {
        set.clear();
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0519_Random_Flip_Matrix.cc
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
 public:
  Solution(int m, int n) {
    row = m;
    col = n;
    size = row * col;
  }
    
  vector<int> flip() {
    while (true) {
      int val = rand() % (row * col);
      if (!flipped.count(val)) {
        flipped.insert(val);
        return {val / col, val % col};
      }
    }
  }
    
  void reset() {
    flipped.clear();
  }

 private:
  int row, col, size;
  unordered_set<int> flipped;
};
